packages/SwingSet/src/kernel/kernel.js

```diff
@@ -286,21 +286,62 @@ export default function buildKernel(kernelEndowments, kernelOptions = {}) {
   function removeVatManager(vatID, reason) {
     const old = ephemeral.vats.get(vatID);
     ephemeral.vats.delete(vatID);
-    const err = reason ? Error(reason) : undefined;
-    old.notifyTermination(err); // XXX TODO: most of the places where notifyTermination gets passed don't need it
+    old.notifyTermination(reason); // XXX TODO: most of the places where notifyTermination gets passed don't need it
     return old.manager.shutdown();
   }
 
-  function terminateVat(vatID, reason) {
+  function removeVatManagerOnFailure(vatID, reason) {
```

Comment:
Elsewhere we seem to have a `terminate(reason)`/`terminateOnFailure(error)` pattern going on, so I think this should be `removeVatManagerOnFailure(error)`. It's an odd factoring here, though, I don't like the duplication of the delete/shutdown calls, but I'm not sure I'd prefer a single function with an extra argument to distinguish between the error/non-error cases.

Reply:
JS already biases us towards the dual method style (Promise resolve vs reject, generator return vs throw) and we have already been following it (notifier finish vs fail, seat exit vs kickOut). I think we should follow it consistently, with more predictable naming, and extend it to zcf shutdown and vat termination, as @Chris-Hibbert is doing in this PR.

However, seeing 
> terminate(reason)/terminateOnFailure(error)
I think my previous advice was misunderstood. The terminal success value should never be called `reason`. The parameter `reason` in JS is used exclusive for the terminal failure value, which is almost always an error. For example, we often see code like
```js
fooP.then(value => stuff..., reason => stuff...)
```
where `reason` is the value with which the promise is rejected. As the name implies, it normally represents the alleged reason why the promise is rejected.

If it were necessarily always an error, then `error` would be a fine name instead of `reason`. However, it is not necessarily an error unless we enforce that it is, which I do not suggest. It's actually impossible today to reliably detect, in portable JS, whether an object is an error. (Which is causing me some grief btw) So please call the terminal failure value `reason`, pass only errors as reasons if you originate the reason, but don't assume the reason must be an error if you receive one, like in the above promise code.

Please don't call anything else, especially the terminal success value, `reason`.

Reply:
When I was implementing the vat termination stuff originally, the distinction I would make in my discussions with Brian was between what I (somewhat jokingly) called "termination for cause" vs. "peremptory termination".  One is justified by some explicit reason why the vat just _has_ to die and the other is just because I say so.  This is somewhat different from the JS resolve/reject distinction, because both kinds of vat termination represent an abrupt cessation of activity, possibly with unfinished business remaining (in the form of unresolved promises and messages to the vat still waiting in the kernel run queue) and possibly with external relationships that may be left in an inconsistent state.  In no sense is either kind of termination some kind of fulfillment event.  As the API was originally designed, terminations for cause originate with the kernel because in its supervision of the vat it detects that something bad has happened (meter exhaustion, bad c-list juju, whatever), whereas peremptory terminations originate with the holder of the vat admin facet, and are explicit acts based (presumably) on use-case-specific application logic that the kernel knows nothing about but, very importantly, made by a party outside the vat being killed who has no direct visibility into what's going on inside the vat being killed.

We had been distinguishing the two by rejecting or resolving the `done` promise, but the resolution path conveyed no information aside from the fact of termination.  The concept was that a rejection was like an error (which it was) whereas a resolution was a signal that the termination had completed successfully -- but it's the act of termination that was successful, not the vat.  The vat may still be a complete trainwreck.

It seems plausible to me to add, as an aid to coordination, a way for the party initiating a peremptory termination (i.e., the holder of the vat admin facet) to include some kind of annotation to be passed along to whoever is watching (i.e., any holders of the `done` promise) so that they can have some indication why the admin pulled the plug.  While I understand Mark's desire to avoid using the word "reason" for this (and we can certainly avoiding naming the value this in our code) I don't seem any way to readily discourage people from using that word in casual conversation since a reason is exactly what it is.

It does seem wrong to me, however, to have the description of why termination happened (whatever we name it) to get mixed up in the choice of fulfilling or rejecting the `done` promise.  That's crossing the streams in a bad way.

Reply:
@FUDCo, your two kinds of termination don't seem to include "everything is done, this contract isn't needed anymore." 

In this case, I expect the clean-up of outstanding promises to not find anything, and if it does find some unfulfilled promises, I want to be notified that the attempt to terminate successfully was unsuccessful. If nothing abnormal is found, the holder of the `done` promise should just be told that it fulfilled.

Do you see this as like one of your two types? From my point of view, I expect the other two to follow one rejection path, but possibly with different messages.

Reply:
Basically, what I called "peremptory termination" is a matter between the vat admin and whoever it gave the `done` promise to.  That's why I think having an information pathway between them makes sense for some use cases.  What I don't want to do is comingle signals from the kernel with signals from the vat admin.

Reply:
Having looked through the code in the PR more closely, the thing that still bugs me (aside from minor structural issues that I flagged in my review comments) is that the sole difference between `terminate` and `terminateOnFailure` is how the signal is received on the `done` end, i.e., as fulfill or reject.  In particular, by tying `terminate` to fulfill and `terminateOnFailure` to reject, the connotation is that with `terminateOnFailure` there has been some kind of error condition that killed things in the midst of processing (which is true generally speaking), but that on plain `terminate` it was a normal exit and everything is fine.  But there is nothing in the kernel's handling of termination that realizes this distinction; from the kernel & vats' perspective, these two calls do _exactly_ the same thing: abruptly kill the vat, throwing away any work for it that was in the queue and severing any dangling promises it was the decider for.  It's possible that the invoker of `terminate` knows that there's no important work in the pipeline and no important unresolved promises still pending, but that's application-space knowledge that is not captured anywhere and so there is nothing the kernel can do to enforce this.

Perhaps, in an ideal world, if there are outstanding messages in the queue or pending promises, `terminate` would refuse to terminate and instead throw an error back to its caller, but because of distributed communication lags there's no way for the kernel to tell that there aren't messages in flight to the vat that it just hasn't seen yet.

An approach that would actually be feasible (and would address most of my concerns here) would be for `terminate` to send some kind of signal to the vat asking it to voluntarily terminate itself if it thought its state was OK for that and refuse/throw/reject if not.  So here's a strawman API to consider: instead of giving the vat admin the option to send `terminate` vs. `terminateOnFailure`, instead have the vat admin only be able to send an involuntary terminate that always rejects the `done` promise, and have the vat itself be able to shut itself down (presumably in response to an ordinary message) via a syscall that would fulfill the `done` promise (in which case it might make sense to also allow the vat to shut itself down with a failure/reject mode as well, since it might be able to detect unrecoverable internal error conditions on its own -- in fact, the more I think about this, this is probably something we want regardless).

Reply:
Huh, that's an interesting feature request: when shutting down a vat, reveal whether there were outstanding unresolved promises or not (i.e. any promise for which the late vat was the decider), via the fulfill/reject resolution of the admin-facet's `done` promise. Basically a deliberate `terminate()` call is deemed to be "successful" (and causes done.then(fulfillCB)` to fire) iff the vat left no orphaned promises, otherwise the terminate is deemed "failure" (and causes done.catch(rejectCB)` to fire). Vats which die because of metering faults also cause `done` to reject. Among those three cases, in the last two cases the reject option is selected by the kernel, but the timing of the first two cases is decided by the userspace `terminate()` call, whereas the timing of the last case is decided by the kernel (well, by the doomed vat when it misbehaves somehow).

I can imagine distinction being interesting from userspace, although I can't readily imagine what it would do with the information.

I think I'd be nervous to provide that distinction from the kernel. It would reveal information about how far the kernel's run-queue has gotten before the `terminate` was processed: messages which cause/allow the doomed vat to resolve some outstanding result promise may or may not have been delivered by the time shutdown occurs. That exposes behavior of the kernel scheduler which we mostly haven't exposed before, and I'm concerned about what constraints that puts on us in the future. The less information we reveal, the more freedom we retain to upgrade/improve/fix the kernel.

Also, messages which are aimed at the doomed vat, but which haven't been delivered yet, will have their own result promises. We currently reject these later, when it gets to the top of the queue, and the message then "splats" against the now-closed door of the missing vat. Would we want these outstanding result promises to qualify as the terminate "failing"? We can't easily tell that a given message in the kernel run-queue is destined for any given vat: those messages are aimed at an object (or a promise), and the vat which owns it might change at any point up until the delivery occurs. So to include these pending messages in the "termination failure predicate" might require delaying the notification until the run-queue had been flushed, which doesn't sound great.

Reply:
This is why I'm increasingly liking the concept that involuntary shutdown (triggered from outside the vat, by the kernel or the vat admin) _always_ causes a rejection, while voluntary shutdown (triggered from inside the vat, by the vat's own code) may reject or fulfill at the vat's discretion.  The termination logic in the kernel always rejects, unconditionally, all unresolved promises and any messages delivered after termination.  This is clean, simple, easy to understand, and doesn't fall into the trap of encouraging the holder of the vat admin facet to think everything is fine when it might not be.  A voluntary shutdown with outstanding promises is the vat saying, in effect, "yes, there may be outstanding promises but as far as I'm concerned that's OK".